### PR TITLE
Update ethereum/tests fixtures to v6.0.0-beta.2 + fix revealed consensus failures

### PR DESCRIPTION
**Pull request by veox**
### What was wrong?

`py-evm` doesn't yet pass the full Constantinople test suite.

This started as a copy of PR #1181, so that Christoph doesn't have to act as a gatekeeper for every tiny fix I propose.

Closes #1181 as superseded/conflicting.

### How was it fixed?

The PR updates the upstream `ethereum/tests` (git sub-module in dir `fixtures`), enables tests for Constantinople, and fixes the breaks introduced by that.

NOTE: some test are very slow or consume a lot of memory, making it impossible to pass CI on CircleCI's "open source contributor" tier (the process gets OOM-killed). All of these are in `SLOW_TESTS`. All tests pass locally for me.

### Cute Animal Picture

put a cute animal picture link inside the parentheses

Source: [Valdas Augustinas @ efoto.lt](https://www.efoto.lt/node/404410)

### CODE REVIEW

**Comment by lithp** [comment]
I wonder if a different workflow might make developing easier and give us slightly better PRs.

The current process seems to be: pick one of the failing tests here and try to fix the problem, once you fix it open a PR against master. Once it's merged rebase this branch and repeat.

What if instead we: add an equivalent of `SLOWEST_TESTS`, maybe it's called `KNOWN_FAILURES'. skip all tests in KNOWN_FAILURES and merge this PR. Then, future PRs can make changes and simultaneously remove the tests they fix from the KNOWN_FAILURES set.

This has some benefits:
- PRs automatically document which tests they were designed to fix. This makes git blame a lot more useful. I think I remember @carver wanting this.
- No rebasing or playing with branches. Just make a PR, merge it, and you're done.

I don't feel very strongly about this so I'm happy to be shot down, but I can add a commit with KNOWN_FAILURES to this PR if that sounds reasonable.

**Comment by veox** [comment]
@lithp I'm OK with that approach, too, especially since I'm already sitting on lists of failing tests (didn't share because extra effort).

Please don't push anything to a branch marked `[WIP]`.

The change can be done more easily when PR #1577 is merged, so that `KNOWN_FAILURES` doesn't have to be repeated across both "blockchain" and "state" upstream tests.

**Comment by lithp** [comment]
Sounds great, tomorrow morning I'll open a new PR based on #1577. Good call on using xfail instead!

**Comment by cburgdorf** [comment]
@lithp I like your proposed approach. The biggest risk with that that I see is that we forget a bout the known failures. But I one possible way to deal with that would be to just keep one PR open that removes everything from `KNOWN_FAILURES` (so CI will fail) and keep it open until we have worked out all of the cases.

At least to me an open PR is the strongest reminder of things that should not be forgotten.

**Comment by veox** [comment]
Regarding failures: it seems that most upstream tests generated from `InitCollisionFiller.json` fail in the same way that the tests already in `INCORRECT_UPSTREAM_TESTS` do, due to the "synthetic" state:

https://github.com/ethereum/tests/blob/v6.0.0-beta.2/src/GeneralStateTestsFiller/stSStoreTest/InitCollisionFiller.json#L117-L134

This results in a gas use difference in all four cases generated from that filler.

Only one of the four has a `REVERT`; the feature being tested here is gas use on collision.

-----

Since there's also a lot of failing tests for `SSTORE` net gas metering, I'll refrain from calling the above "incorrect upstream" or "known failure".

**Comment by veox** [comment]
Note that, according to [the list](https://gist.github.com/veox/d2a07cfe86fdce11967dd81489c65899), there are only three classes of failures currently:

* `InitCollision` (discussed above);
* [EIP-1283](https://eips.ethereum.org/EIPS/eip-1283) tests, tracked upstream in https://github.com/ethereum/tests/issues/483;
* 9 of `bcInvalidHeaderTest`s (for all chains, not just `Constantinople`).

**Comment by veox** [comment]
From tracing it seems that `Constantinople` takes no heed of `STATICCALL` locking state.

E.g., running

```pytest tests/json-fixtures/test_blockchain.py -k 'test_blockchain_fixtures[/home/veox/src/py-evm/fixtures/BlockchainTests/GeneralStateTests/stStaticCall/static_callBasic_d0g0v0.json:static_callBasic_d0g0v0_Constantinople]'
```

results in:

```
   TRACE  12-12 16:55:30            logging.py  OPCODE: 0xfa (STATICCALL) | pc: 15
   TRACE  12-12 16:55:30            logging.py  GAS CONSUMPTION: 978723 - 700 -> 978023 (STATICCALL)
   TRACE  12-12 16:55:30            logging.py  MEMORY: size (0 -> 0) | cost (0 -> 0)
   TRACE  12-12 16:55:30            logging.py  MEMORY: size (0 -> 0) | cost (0 -> 0)
   TRACE  12-12 16:55:30            logging.py  GAS CONSUMPTION: 978023 - 100000 -> 878023 (STATICCALL)
   TRACE  12-12 16:55:30            logging.py  COMPUTATION STARTING: gas: 100000 | from: 0x1000000000000000000000000000000000000000 | to: 0x1000000000000000000000000000000000000001 | value: 0 | depth 1 | static: y
   TRACE  12-12 16:55:30            logging.py  OPCODE: 0x60 (PUSH1) | pc: 0
   TRACE  12-12 16:55:30            logging.py  GAS CONSUMPTION: 100000 - 3 -> 99997 (PUSH1)
   TRACE  12-12 16:55:30            logging.py  OPCODE: 0x60 (PUSH1) | pc: 2
   TRACE  12-12 16:55:30            logging.py  GAS CONSUMPTION: 99997 - 3 -> 99994 (PUSH1)
   TRACE  12-12 16:55:30            logging.py  OPCODE: 0x55 (SSTORE) | pc: 4
   TRACE  12-12 16:55:30            logging.py  GAS CONSUMPTION: 99994 - 200 -> 99794 (SSTORE: 0x1000000000000000000000000000000000000001[1] -> 1 (current: 1 / original: 1))
   TRACE  12-12 16:55:30            logging.py  OPCODE: 0x0 (STOP) | pc: 4
   TRACE  12-12 16:55:30            logging.py  COMPUTATION SUCCESS: from: 0x1000000000000000000000000000000000000000 | to: 0x1000000000000000000000000000000000000001 | value: 0 | depth: 1 | static: y | gas-used: 206 | gas-remaining: 99794
```

It should have errored out on trying to `SSTORE` within a `STATICCALL` frame.

As mentioned before, `Byzantium` acts as expected in this same test.

**Comment by veox** [comment]
... and that magically fixed every remaining "state" failure, apart from `InitCollision`.

The remaining failures that I'm experiencing locally are actually not tested for on `master` or in this PR, so aren't revealed by CI.

Currently waiting on PR #1577 to get merged, so this PR can get re-worked without bothering with the "state" branch of tests.

**Comment by pipermerriam** [comment]
@lithp I'd be fine with your `KNOWN_FAILURES` approach if it was a dedicated branch that we merged fixes into **and then** once things are all passing we merge into master, clearing out the `KNOWN_FAILURES` set.  I'm uncomfortable doing it in `master` for the reason @cburgdorf described.  I don't want to forget about them.

**Comment by veox** [comment]
@pipermerriam I'm now reasonably certain it will not come to `KNOWN_FAILURES` at all.

There's only one class of failure that I haven't looked at or fixed (locally), the `bcInvalidHeaderTest`s.

I'm not working at getting CI green here, because this work would be overridden by PR #1577 anyway.

**Comment by pipermerriam** [comment]
:heart: @veox

File: eth/vm/forks/constantinople/opcodes.py [hunk 1]
```diff
@@ -11,7 +11,8 @@
     opcode_values,
 )
 from eth.vm.forks.byzantium.opcodes import (
-    BYZANTIUM_OPCODES
+    BYZANTIUM_OPCODES,
+    ensure_no_static
```
Comment (veox) [inline eth/vm/forks/constantinople/opcodes.py]: Could possibly be moved "out" of any fork-specific file.

File: tests/json-fixtures/test_blockchain.py [hunk 2]
```diff
@@ -45,6 +45,11 @@
     # The result is in conflict with the yellow-paper:
     # * https://github.com/ethereum/py-evm/pull/1224#issuecomment-418800369
     ('GeneralStateTests/stRevertTest/RevertInCreateInInit_d0g0v0.json', 'RevertInCreateInInit_d0g0v0_Byzantium'),  # noqa: E501
+    ('GeneralStateTests/stRevertTest/RevertInCreateInInit_d0g0v0.json', 'RevertInCreateInInit_d0g0v0_Constantinople'),  # noqa: E501
+    # The CREATE2 variant seems to have been derived from the one above - it, too,
+    # has a "synthetic" state, on which py-evm flips.
+    # * https://github.com/ethereum/py-evm/pull/1181#issuecomment-446330609
+    ('GeneralStateTests/stCreate2/RevertInCreateInInitCreate2_d0g0v0.json', 'RevertInCreateInInitCreate2_d0g0v0_Constantinople'),  # noqa: E501
```
Comment (veox) [inline tests/json-fixtures/test_blockchain.py]: This change is made obsolete by PR #1577. Will rebase and rework into `BlockchainTests` when that is merged.

File: tests/json-fixtures/test_state.py [hunk 3]
```diff
@@ -144,6 +146,11 @@ def expand_fixtures_forks(all_fixtures):
     # The result is in conflict with the yellow-paper:
     # * https://github.com/ethereum/py-evm/pull/1224#issuecomment-418800369
     ('stRevertTest/RevertInCreateInInit.json', 'RevertInCreateInInit', 'Byzantium', 0),
+    ('stRevertTest/RevertInCreateInInit.json', 'RevertInCreateInInit', 'Constantinople', 0),
+    # The CREATE2 variant seems to have been derived from the one above - it, too,
+    # has a "synthetic" state, on which py-evm flips.
+    # * https://github.com/ethereum/py-evm/pull/1181#issuecomment-446330609
+    ('stCreate2/RevertInCreateInInitCreate2.json', 'RevertInCreateInInitCreate2', 'Constantinople', 0),  # noqa: E501
```
Comment (veox) [inline tests/json-fixtures/test_state.py]: This change is made obsolete by PR #1577. Will rebase and rework into `BlockchainTests` when that is merged.

File: tox.ini [hunk 4]
```diff
@@ -3,7 +3,7 @@ envlist=
     py{35,36}-{core,database,transactions,vm,native-blockchain}
     py{36}-{benchmark,p2p,trinity,lightchain_integration,beacon}
     py{36}-rpc-blockchain
-    py{36}-rpc-state-{frontier,homestead,eip150,eip158,byzantium,quadratic}
+    py{36}-rpc-state-{frontier,homestead,eip150,eip158,byzantium,constantinople,quadratic}
```
Comment (veox) [inline tox.ini]: This change is made obsolete by PR #1577. Will rebase when that is merged.

**Comment by veox** [comment]
Rebased branch on latest `master` and reworked it; old variant backed up in https://github.com/veox/py-evm/tree/update-fixtures-to-v6.0.0-beta.2-old.

File: tests/json-fixtures/test_blockchain.py [hunk 5]
```diff
@@ -37,80 +37,109 @@
...
+    ('GeneralStateTests/stStaticCall/static_Call1MB1024Calldepth_d1g0v0.json', 'static_Call1MB1024Calldepth_d1g0v0_Constantinople'),  # noqa: E501
+    ('GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024_d0g0v0.json', 'Delegatecall1024_d0g0v0_Constantinople'),  # noqa: E501
+    ('GeneralStateTests/stChangedEIP150/Callcode1024BalanceTooLow_d0g0v0.json', 'Callcode1024BalanceTooLow_d0g0v0_Constantinople'),  # noqa: E501
+    ('GeneralStateTests/stDelegatecallTestHomestead/CallRecursiveBombPreCall_d0g0v0.json', 'CallRecursiveBombPreCall_d0g0v0_Constantinople'),  # noqa: E501
+    ('GeneralStateTests/stDelegatecallTestHomestead/Call1024BalanceTooLow_d0g0v0.json', 'Call1024BalanceTooLow_d0g0v0_Constantinople'),  # noqa: E501
+    ('GeneralStateTests/stRevertTest/LoopCallsDepthThenRevert3_d0g0v0.json', 'LoopCallsDepthThenRevert3_d0g0v0_Constantinople'),  # noqa: E501
+    ('GeneralStateTests/stCallCreateCallCodeTest/CallRecursiveBombPreCall_d0g0v0.json', 'CallRecursiveBombPreCall_d0g0v0_Constantinople'),  # noqa: E501
+    ('bcForkStressTest/ForkStressTest.json', 'ForkStressTest_Constantinople'),  # noqa: E501
```
Comment (veox) [inline tests/json-fixtures/test_blockchain.py]: Perhaps should move `bcForkStressTest` into `blockchain_fixture_mark_fn()`?..

File: tests/json-fixtures/test_blockchain.py [hunk 6]
```diff
@@ -37,80 +37,98 @@
...
-    ('GeneralStateTests/stSystemOperationsTest/CallRecursiveBombLog_d0g0v0.json', 'CallRecursiveBombLog_d0g0v0_Homestead'),  # noqa: E501
-    ('GeneralStateTests/stSystemOperationsTest/CallRecursiveBomb1_d0g0v0.json', 'CallRecursiveBomb1_d0g0v0_Frontier'),  # noqa: E501
-    ('GeneralStateTests/stSystemOperationsTest/CallRecursiveBombLog2_d0g0v0.json', 'CallRecursiveBombLog2_d0g0v0_Homestead'),  # noqa: E501
-    ('GeneralStateTests/stSystemOperationsTest/CallRecursiveBomb0_OOG_atMaxCallDepth_d0g0v0.json', 'CallRecursiveBomb0_OOG_atMaxCallDepth_d0g0v0_Frontier'),  # noqa: E501
-    ('GeneralStateTests/stCallCreateCallCodeTest/Call1024BalanceTooLow_d0g0v0.json', 'Call1024BalanceTooLow_d0g0v0_Byzantium'),  # noqa: E501
+    ('GeneralStateTests/stChangedEIP150/Call1024BalanceTooLow_d0g0v0.json', 'Call1024BalanceTooLow_d0g0v0_Constantinople'),  # noqa: E501
     ('GeneralStateTests/stChangedEIP150/Call1024BalanceTooLow_d0g0v0.json', 'Call1024BalanceTooLow_d0g0v0_EIP150'),  # noqa: E501
-    ('bcForkStressTest/ForkStressTest.json', 'ForkStressTest_EIP150'),  # noqa: E501
```
Comment (veox) [inline tests/json-fixtures/test_blockchain.py]: This is now disabled as a class, see `blockchain_fixture_mark_fn()` below.

File: tox.ini [hunk 7]
```diff
@@ -32,6 +32,8 @@ commands=
     beacon: pytest {posargs:tests/beacon/}
     # The following test seems to consume a lot of memory. Restricting to 3 processes reduces crashes
     rpc-state-byzantium: pytest -n3 {posargs:tests/trinity/json-fixtures-over-rpc/test_rpc_fixtures.py -k 'GeneralStateTests and not stQuadraticComplexityTest and Byzantium'}
+    # Uncomment the next line + modify test_rpc_fixtures.py when Constantinople is included in the mainnet config
```
Comment (carver) [inline tox.ini]: Old comment can be deleted?

```suggestion
```

Comment (carver) [reply tox.ini]: We can do it in a follow-up. I don't want to do it myself in case I misread the comment.

Comment (veox) [reply tox.ini]: This was an oversight. It shouldn't have been there - oops.

Comment (carver) [reply tox.ini]: Kinda figured, but also didn't want to hold up the release over it.

**Pull request merged**